### PR TITLE
Implement lazy properties

### PR DESCRIPTION
**Pull request by pawelpasterz**
This will fix https://github.com/runningcode/fladle/issues/92

What was done?
1) Implement all remaining properties to use lazy API.
2) small refactor

What was not done?
1) there is a test failing (locally) `ConfigurationCacheTest#runFlank` (it is also falling on the master branch)
2) update with latest flank's command
3) enable to run robo test without robo script
4) refactor of `YamlWriterTest`

I decided to left above as they are since the scope of this PR is rather big

### CODE REVIEW

**Comment by pawelpasterz** [comment]
Hey, @runningcode please take a look, thanks!

File: build.gradle [hunk 1]
```diff
@@ -16,6 +16,7 @@ plugins {
     id "com.github.ben-manes.versions" version "0.30.0"
     id "com.osacky.fulladle"
     id "org.jmailen.kotlinter" version "3.0.2"
+    id "org.jetbrains.kotlin.jvm" version "$kotlin_version"
```
Comment (runningcode) [inline build.gradle:19]: can you tell me more about this change? why do we need this?

File: buildSrc/src/main/java/com/osacky/flank/gradle/FladleConfig.kt [hunk 2]
```diff
@@ -154,22 +159,23 @@ interface FladleConfig {
...
+  val testTargetsAlwaysRun: ListProperty<String>
 
   /**
    * A list of device-path: file-path pairs that indicate the device paths to push files to the device before starting tests, and the paths of files to push.
    * Device paths must be under absolute, whitelisted paths (${EXTERNAL_STORAGE}, or ${ANDROID_DATA}/local/tmp).
    * Source file paths may be in the local filesystem or in Google Cloud Storage (gs://…).
    */
   @get:Input
-  var otherFiles: Map<String, String>
+  @get:Optional
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/FladleConfig.kt:177]: why did we add optional here, but not in other properties?

File: build.gradle (hunk 1 above)
Comment (pawelpasterz) [reply build.gradle:19]: Yeah, that was one of my experiments and it was left unintentionally 👍  Thanks

File: buildSrc/src/main/java/com/osacky/flank/gradle/FladleConfig.kt (hunk 2 above)
Comment (pawelpasterz) [reply buildSrc/src/main/java/com/osacky/flank/gradle/FladleConfig.kt:177]: Nice catch will remove that

File: sample/build.gradle [hunk 3]
```diff
@@ -36,18 +36,18 @@ fladle {
     smartFlankGcsPath = "gs://test-lab-yr9w6qsdvy45q-iurp80dm95h8a/flank/test_app_android.xml"
     configs {
         oranges {
-            useOrchestrator = false
-            testTargets = [
+            useOrchestrator.set(false)
```
Comment (runningcode) [inline sample/build.gradle:39]: this change should not be necesary. If you revert, does it still work? I would like to keep backward compatibility with the groovy dsl.

File: sample/build.gradle [hunk 4]
```diff
@@ -36,18 +36,18 @@ fladle {
...
+            testTargets.set(project.provider { [
```
Comment (runningcode) [inline sample/build.gradle:40]: same here as above and for the rest of the file.

File: docs/configuration.md [hunk 5]
```diff
@@ -76,10 +76,10 @@ fladle {
     smartFlankGcsPath = "gs://tmp_flank/flank/test_app_android.xml"
     configs {
         oranges {
-            useOrchestrator = false
-            testTargets = [
+            useOrchestrator.set(false)
```
Comment (runningcode) [inline docs/configuration.md:79]: thanks for updating the documentation. I believe that this change is not necessary, but we should make this change in the kotlin example in the docs.

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt [hunk 6]
```diff
@@ -27,167 +29,92 @@ internal class YamlWriter {
...
+      prop.get().forEach { custom(it) }
     }
   }
 
-  private fun flakyTestAttemptsLine(flakyTestAttempts: Int): String {
-    val label = "num-flaky-test-attempts"
-    return "  $label: $flakyTestAttempts"
-  }
+  private val <T> ListProperty<T>.isPresentAndNotEmpty
+    get() = isPresent && get().isEmpty().not()
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:114]: Is it possible to write it like so?
```suggestion
    get() = isPresent && get().isNotEmpty()
```

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt [hunk 7]
```diff
@@ -27,167 +29,92 @@ internal class YamlWriter {
...
 
-  private fun repeatTestsLine(repeatTests: Int): String {
-    val label = "num-test-runs"
-    return "  $label: $repeatTests"
-  }
+  private val <T, K> MapProperty<T, K>.isPresentAndNotEmpty
+    get() = isPresent && get().isEmpty().not()
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:117]: same suggestion here:
```suggestion
    get() = isPresent && get().isNotEmpty()
```

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt [hunk 8]
```diff
@@ -1,23 +1,25 @@
...
 internal class YamlWriter {
 
   internal fun createConfigProps(config: FladleConfig, base: FlankGradleExtension): String {
-    if (base.projectId == null) {
+    if (base.projectId.orNull == null) {
       check(base.serviceAccountCredentials.isPresent) { "ServiceAccountCredentials in fladle extension not set. https://github.com/runningcode/fladle#serviceaccountcredentials" }
     }
     check(base.debugApk.isPresent) { "debugApk must be specified" }
-    check(base.instrumentationApk.isPresent xor !base.roboScript.isNullOrBlank()) {
+    check(base.instrumentationApk.isPresent xor base.roboScript.isPresent) {
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:15]: the previous logic checked if the `roboScript` was not null and not blank. Should we check that here as well?

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt [hunk 9]
```diff
@@ -27,167 +29,92 @@ internal class YamlWriter {
       if (base.instrumentationApk.isPresent) {
         appendln("  test: ${base.instrumentationApk.get()}")
       }
-      appendln(deviceString)
+      if (config.devices.isPresentAndNotEmpty) appendln(createDeviceString(config.devices.get()))
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:32]: Is the device a required property in flank? What would happen if it were omitted?

I believe the previous logic would alway add it, this logic only add it if it is present. Why the change?

**Review by runningcode — CHANGES_REQUESTED** [review]
Looking good. just some small logic changes I noticed. I haven't finished my review though.

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt (hunk 9 above)
Comment (pawelpasterz) [reply buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:32]: Actually not, we can omit devices and `NexusLowRes` will be taken as default. TBH the only required properties (from flank's perspective) is app apk and test apk. I have in my mind to change it in fladle but in separate PR

File: sample/build.gradle (hunk 3 above)
Comment (pawelpasterz) [reply sample/build.gradle:39]: I am afraid it is necessary. Without it, tests are failing. What is more, if you add `fullJunitResult` (which on current master branch is `Property<Boolean>`) to `sample` `build.gradle` like:
```gradle
fladle {
    ...
    configs {
        oranges {
            ...
            flakyTestAttempts = 3
            fullJunitResult = true
        }
        ...
    }
    ...
}
```
it fails with:
```
FAILURE: Build failed with an exception.

* Where:
Build file '/Users/gogo/fladle/sample/build.gradle' line: 44

* What went wrong:
A problem occurred evaluating project ':sample'.
> Cannot cast object 'true' with class 'java.lang.Boolean' to class 'org.gradle.api.provider.Property'
```

File: docs/configuration.md (hunk 5 above)
Comment (pawelpasterz) [reply docs/configuration.md:79]: https://github.com/runningcode/fladle/pull/176#discussion_r493391055

File: sample/build.gradle (hunk 4 above)
Comment (pawelpasterz) [reply sample/build.gradle:40]: https://github.com/runningcode/fladle/pull/176#discussion_r493391055

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt (hunk 8 above)
Comment (pawelpasterz) [reply buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:15]: TBH I based my implementation on `instrumentationApk` check, but yeah, for now we should check for path being blank as well.

FYI, this will change with https://github.com/runningcode/fladle/issues/165

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt (hunk 9 above)
Comment (pawelpasterz) [reply buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:32]: But I can change implementation to reflect the previous one, and change it with another PR

File: buildSrc/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt [hunk 10]
```diff
@@ -55,51 +57,51 @@ open class FlankGradleExtension @Inject constructor(objects: ObjectFactory) : Fl
...
 
-  override var resultsDir: String? = null
+  override val resultsDir: Property<String> = objects.property()
 
   override val additionalTestApks: ListProperty<String> = objects.listProperty()
 
   override val runTimeout: Property<String> = objects.property()
 
-  override val ignoreFailedTests: Property<Boolean> = objects.property()
+  override val ignoreFailedTests: Property<Boolean> = objects.property<Boolean>().convention(false)
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt:80]: why was this changed to false here?

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt (hunk 9 above)
Comment (runningcode) [reply buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:32]: Makes sense. What version of the API/Android SDK does Flank use? I want to ensure that there are no unexpected changes introduced with this change.

**Review by runningcode — CHANGES_REQUESTED** [review]
Two more questions.
There are a lot of large customers of fladle, and I just want to make sure not to introduce any unexpected changes.

File: buildSrc/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt (hunk 10 above)
Comment (pawelpasterz) [reply buildSrc/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt:80]: Oh...again, unintentionally....There were lots of changes and got a bit lost 😅  Will fix that.

Comment (pawelpasterz) [reply buildSrc/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt:80]: BTW, this is the default value in flank

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt (hunk 9 above)
Comment (pawelpasterz) [reply buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:32]: > Makes sense. What version of the API/Android SDK does Flank use? I want to ensure that there are no unexpected changes introduced with this change.

28

https://github.com/Flank/flank/blob/master/test_runner/src/main/kotlin/ftl/config/FtlConstants.kt#L47-L48

File: buildSrc/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt (hunk 10 above)
Comment (pawelpasterz) [reply buildSrc/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt:80]: Yeah, now I remember why I did it like this

https://github.com/runningcode/fladle/blob/7678e819b1b595b113c93e44d22a2b2cca4ca82d/buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt#L44

I just moved the default value

File: buildSrc/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt [hunk 11]
```diff
@@ -119,7 +119,7 @@ class FladlePluginDelegate {
               project.log("Configuring fladle.debugApk from variant ${this@app.name}")
               extension.debugApk.set(this@app.outputFile.absolutePath)
             }
-            if (extension.roboScript == null && !extension.instrumentationApk.isPresent) {
+            if (!extension.roboScript.isPresent && !extension.instrumentationApk.isPresent) {
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt:122]: in order to maintain the same logic as before, i think this should be:
```suggestion
            if (extension.roboScript.orNull == null && !extension.instrumentationApk.isPresent) {
```

File: buildSrc/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt [hunk 12]
```diff
@@ -20,25 +22,25 @@ open class FlankGradleExtension @Inject constructor(objects: ObjectFactory) : Fl
...
   // Project id is automatically discovered by default. Use this to override the project id.
-  override var projectId: String? = null
+  override val projectId: Property<String> = objects.property()
   override val serviceAccountCredentials: RegularFileProperty = objects.fileProperty()
-  override var useOrchestrator: Boolean = false
-  override var autoGoogleLogin: Boolean = false
-  override var devices: List<Map<String, String>> = listOf(mapOf("model" to "NexusLowRes", "version" to "28"))
+  override val useOrchestrator: Property<Boolean> = objects.property<Boolean>().convention(false)
+  override val autoGoogleLogin: Property<Boolean> = objects.property<Boolean>().convention(false)
+  override val devices: ListProperty<Map<String, String>> = objects.listProperty<Map<String, String>>().convention(listOf(mapOf("model" to "NexusLowRes", "version" to "28")))
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt:29]: thanks for keeping this here. perhaps we can remove it in a separate PR.

File: buildSrc/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt (hunk 10 above)
Comment (runningcode) [reply buildSrc/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt:80]: Ah that makes sense. I like keeping all the default values here in one place. 👍

File: buildSrc/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt [hunk 13]
```diff
@@ -111,41 +113,41 @@ open class FlankGradleExtension @Inject constructor(objects: ObjectFactory) : Fl
...
-      flakyTestAttempts = flakyTestAttempts,
-      directoriesToPull = directoriesToPull,
-      filesToDownload = filesToDownload,
-      environmentVariables = environmentVariables,
-      recordVideo = recordVideo,
-      performanceMetrics = performanceMetrics,
-      resultsBucket = resultsBucket,
-      keepFilePath = keepFilePath,
-      resultsDir = resultsDir,
+      useOrchestrator = objects.property<Boolean>().convention(useOrchestrator),
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt:118]: thanks for changing all of these! 👍

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt [hunk 14]
```diff
@@ -1,23 +1,25 @@
...
 internal class YamlWriter {
 
   internal fun createConfigProps(config: FladleConfig, base: FlankGradleExtension): String {
-    if (base.projectId == null) {
+    if (base.projectId.orNull == null) {
       check(base.serviceAccountCredentials.isPresent) { "ServiceAccountCredentials in fladle extension not set. https://github.com/runningcode/fladle#serviceaccountcredentials" }
     }
     check(base.debugApk.isPresent) { "debugApk must be specified" }
-    check(base.instrumentationApk.isPresent xor !base.roboScript.isNullOrBlank()) {
+    check(base.instrumentationApk.isPresent xor !base.roboScript.orNull.isNullOrBlank()) {
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:15]: 👍

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt (hunk 9 above)
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:32]: i think this `isPresentAndNotEmpty` can be removed now.

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt [hunk 15]
```diff
@@ -27,167 +29,92 @@ internal class YamlWriter {
...
-        appendln("  - class $it")
-      }
-    }
-
-    appendln("  legacy-junit-result: ${config.legacyJunitResult.get()}")
-    appendln("  full-junit-result: ${config.fullJunitResult.get()}")
-    appendln("  output-style: ${config.outputStyle.get()}")
+    appendProperty(config.testShards, name = "max-test-shards")
+    appendProperty(config.shardTime, name = "shard-time")
+    appendProperty(config.repeatTests, name = "num-test-runs")
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:43]: nice cleanup here

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt [hunk 16]
```diff
@@ -27,167 +29,92 @@ internal class YamlWriter {
...
+  }
 
-    if (config.roboDirectives.isNotEmpty()) {
-      appendln("  robo-directives:")
-      config.roboDirectives.forEach {
-        val value = it.getOrElse(2) { "" }.let { stringValue -> if (stringValue.isBlank()) "\"\"" else stringValue }
-        appendln("    ${it[0]}:${it[1]}: $value")
-      }
+  private fun <T> StringBuilder.appendListProperty(
+    prop: ListProperty<T>,
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:103]: In this case, I believe the list property must always be of type string. Can we restrict the API in that way in order to prevent it from being abused or having unintended effects?
```suggestion
  private fun StringBuilder.appendListProperty(
    prop: ListProperty<String>,
```

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt [hunk 17]
```diff
@@ -27,167 +29,92 @@ internal class YamlWriter {
...
-    config.networkProfile?.let {
-      appendln("  network-profile: $it")
-    }
+  private fun <T> StringBuilder.appendProperty(prop: Property<T>, name: String) {
+    if (prop.isPresent) appendln("  $name: ${prop.get()}")
+  }
 
-    config.roboScript?.let {
-      appendln("  robo-script: $it")
+  private fun <T, K> StringBuilder.appendMapProperty(
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:91]: In this case, I believe the map property must always be of type string. Can we restrict the API in that way in order to prevent it from being abused or having unintended effects?
```suggestion
  private fun StringBuilder.appendMapProperty(
```

**Review by runningcode — COMMENTED** [review]
Alright, we're getting pretty close. Sorry its a big PR and its taking me a while to get through. :)
Here's another round of comments.

File: sample/build.gradle (hunk 3 above)
Comment (runningcode) [reply sample/build.gradle:39]: This is bug in Gradle: https://github.com/gradle/gradle/issues/9987 😢

**Review by runningcode — APPROVED** [review]

**Comment by runningcode** [comment]
Thanks @pawelpasterz for doing this! i know it was a lot of work!
